File: learning-python/PySpark/TFRecord/tfrecord_example.py
import os
from pyspark.sql import *
from pyspark.sql.types import *


def main():
    spark = SparkSession.builder \
        .appName("dump_sample") \
        .enableHiveSupport() \
        .getOrCreate()

    job_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../"))
    path = os.path.join(job_path, "data/learning-python/pyspark/tfrecord/tfrecord-example/test-output.tfrecord")

    fields = [StructField("id", IntegerType()), StructField("IntegerCol", IntegerType()),
              StructField("LongCol", LongType()), StructField("FloatCol", FloatType()),
              StructField("DoubleCol", DoubleType()), StructField("VectorCol", ArrayType(DoubleType(), True)),
              StructField("StringCol", StringType())]
    schema = StructType(fields)
    rows = [[11, 1, 23, 10.0, 14.0, [1.0, 2.0], "r1"], [21, 2, 24, 12.0, 15.0, [2.0, 2.0], "r2"]]
    rdd = spark.sparkContext.parallelize(rows)
    df = spark.createDataFrame(rdd, schema)
    df.write.mode("overwrite").format("tfrecord").option("recordType", "Example").save(path)
    df = spark.read.format("tfrecord").option("recordType", "Example").load(path)
    df.show()


if __name__ == "__main__":
    main()

File: learning-spark/src/main/scala/tfrecord/TFRecordExample.scala
package tfrecord

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.{SparkSession, DataFrame, Row}
import org.apache.spark.sql.catalyst.expressions.GenericRow
import org.apache.spark.sql.types._

object TFRecordExample {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
      .setAppName("TFRecordExample")
      .setMaster("local")
    val sc = new SparkContext(conf)
    val spark = SparkSession
      .builder()
      .config(conf)
      .master("local")
      .appName("TFRecordExample")
      .getOrCreate()

    val path = "data/learning-spark/tfrecord-example/test-output.tfrecord"

    val rows: Array[Row] = Array(
      new GenericRow(Array[Any](11, 1, 23L, 10.0F, 14.0, List(1.0, 2.0), "r1")),
      new GenericRow(Array[Any](21, 2, 24L, 12.0F, 15.0, List(2.0, 2.0), "r2")))
    val schema = StructType(List(StructField("id", IntegerType),
      StructField("IntegerCol", IntegerType),
      StructField("LongCol", LongType),
      StructField("FloatCol", FloatType),
      StructField("DoubleCol", DoubleType),
      StructField("VectorCol", ArrayType(DoubleType, true)),
      StructField("StringCol", StringType)))
    val rdd = spark.sparkContext.parallelize(rows)

    //Save DataFrame as TFRecords
    val df: DataFrame = spark.createDataFrame(rdd, schema)
    df.write.format("tfrecord")
      .option("recordType", "Example")
      .option("codec", "org.apache.hadoop.io.compress.GzipCodec")
      .save(path)

    //Read TFRecords into DataFrame.
    //The DataFrame schema is inferred from the TFRecords if no custom schema is provided.
    val importedDf1: DataFrame = spark.read.format("tfrecord").option("recordType", "Example").load(path)
    importedDf1.show()

    //Read TFRecords into DataFrame using custom schema
    val importedDf2: DataFrame = spark.read.format("tfrecord").schema(schema).load(path)
    importedDf2.show()
  }

}
